File: sample/sample-java/SampleMPE.java
//---------------------------------------------------------------------------
//
// TSDuck sample Java application running a chain of plugins
// and handling plugin events in a Java class.
//
//----------------------------------------------------------------------------

import io.tsduck.AbstractPluginEventHandler;
import io.tsduck.AsyncReport;
import io.tsduck.PluginEventContext;
import io.tsduck.TSProcessor;

public class SampleMPE {

    /**
     *  A 32-bit plugin event code ("MPE1" ASCII)
     */
    private static final int EVENT_CODE = 0x4D504531;

    /**
     * A pure Java class which handles TSDuck plugin events.
     */
    private static class EventHandler extends AbstractPluginEventHandler {

        /**
         * This handler is invoked when a plugin signals an event for which this object is registered.
         * @param context An instance of PluginEventContext containing the details of the event.
         * @param data A byte array containing the data of the event.
         */
        @Override
        public boolean handlePluginEvent(PluginEventContext context, byte[] data) {
            System.out.printf("==== Event code: 0x%X from plugin #%d (%s), data size: %d bytes, at TS packet %d\n",
                              context.eventCode(), context.pluginIndex(), context.pluginName(), data.length, context.pluginPackets());
            System.out.printf("MPE datagram: %s\n", SampleUtils.bytesToHex(data));
            return true;
        }
    }

    /**
     * Main program.
     * @param args Command line arguments.
     */
    public static void main(String[] args) {

        // Create a report to log multi-threaded messages.
        AsyncReport rep = new AsyncReport();

        // Create a user-defined event handler to catch plugin events.
        EventHandler handler = new EventHandler();

        // Create a TS processor using the report.
        TSProcessor tsp = new TSProcessor(rep);

        // Register our event handler in the TS processor.
        tsp.registerEventHandler(handler, EVENT_CODE);

        // Set the plugin chain.
        tsp.input = new String[] {"http", "https://github.com/tsduck/tsduck-test/raw/master/input/test-016.ts"};
        tsp.plugins = new String[][] {{"mpe", "--pid", "2001", "--max-datagram", "2", "--event-code", String.valueOf(EVENT_CODE)}};
        tsp.output = new String[] {"drop"};

        // Run the TS processing and wait until completion.
        tsp.start();
        tsp.waitForTermination();
        tsp.delete();

        // Delete the event handler.
        handler.delete();

        // Terminate the asynchronous report.
        rep.terminate();
        rep.delete();
    }
}

File: sample/sample-python/sample-mpe.py
#!/usr/bin/env python
#----------------------------------------------------------------------------
#
# TSDuck sample Python application running a chain of plugins
# and handling plugin events in a Python class.
#
#----------------------------------------------------------------------------

import tsduck

# A 32-bit plugin event code ("MPE1" ASCII)
EVENT_CODE = 0x4D504531

# A pure Python class which handles TSDuck plugin events.
class EventHandler(tsduck.AbstractPluginEventHandler):

    # This method is invoked each time a TSDuck plugin signals an event for which we registered.
    def handlePluginEvent(self, context, data):
        print("==== Event code: 0x%X from plugin #%d (%s), data size: %d bytes, at TS packet %d" %
              (context.event_code, context.plugin_index, context.plugin_name, len(data), context.plugin_packets))
        print("MPE datagram: %s" % (data.hex()))
        return True

# Create an asynchronous report to log multi-threaded messages.
rep = tsduck.AsyncReport()

# Create an event handler to catch plugin events.
handler = EventHandler()

# Create a TS processor, register our event handler, set plugin chain.
tsp = tsduck.TSProcessor(rep)
tsp.registerEventHandler(handler, EVENT_CODE)
tsp.input = ['http', 'https://github.com/tsduck/tsduck-test/raw/master/input/test-016.ts']
tsp.plugins = [
    ['mpe', '--pid', '2001', '--max-datagram', '2', '--event-code', str(EVENT_CODE)],
]
tsp.output = ['drop']

# Run the TS processing and wait until completion.
tsp.start()
tsp.waitForTermination()
tsp.delete()

# Delete the handler.
handler.delete()

# Terminate the asynchronous report.
rep.terminate()
rep.delete()
